contracts/adapters/SwapAdapter.sol

```diff
@@ -0,0 +1,388 @@
+
+/**
+    @title Contract that swaps tokens to ETH or ETH to tokens using Uniswap
+        and then makes a deposit to the Bridge.
+    @author ChainSafe Systems.
+ */
+contract SwapAdapter is AccessControl {
+
+    using SafeERC20 for IERC20;
+
+    IBridge public immutable _bridge;
+    address immutable _weth;
+    IV3SwapRouter public _swapRouter;
+    INativeTokenAdapter _nativeTokenAdapter;
+
+    mapping(address => bytes32) public tokenToResourceID;
+
+    // Used to avoid "stack too deep" error
+    struct LocalVars {
+        bytes32 resourceID;
+        bytes depositDataAfterAmount;
+        uint256 fee;
+        address feeHandlerRouter;
+        uint256 amountOut;
+        uint256 swapAmount;
+        bytes path;
+        IV3SwapRouter.ExactInputParams params;
+        bytes depositData;
+        address ERC20HandlerAddress;
+        uint256 leftover;
+    }
+
+    error CallerNotAdmin();
+    error AlreadySet();
+    error TokenInvalid();
+    error PathInvalid();
+    error MsgValueLowerThanFee(uint256 value);
+    error InsufficientAmount(uint256 amount);
+    error FailedFundsTransfer();
+    error AmountLowerThanFee(uint256 amount);
+
+    event TokenResourceIDSet(address token, bytes32 resourceID);
+    event TokensSwapped(address token, uint256 amountOut);
+
+    constructor(
+        IBridge bridge,
+        address weth,
+        IV3SwapRouter swapRouter,
+        INativeTokenAdapter nativeTokenAdapter
+    ) {
+        _bridge = bridge;
+        _weth = weth;
+        _swapRouter = swapRouter;
+        _nativeTokenAdapter = nativeTokenAdapter;
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    modifier onlyAdmin() {
+        if (!hasRole(DEFAULT_ADMIN_ROLE, msg.sender)) revert CallerNotAdmin();
+        _;
+    }
+
+    // Admin functions
+    function setTokenResourceID(address token, bytes32 resourceID) external onlyAdmin {
+        if (tokenToResourceID[token] == resourceID) revert AlreadySet();
+        tokenToResourceID[token] = resourceID;
+        emit TokenResourceIDSet(token, resourceID);
+    }
+
+    /**
+        @notice Function for depositing tokens, performing swap to ETH and bridging the ETH.
+        @param destinationDomainID  ID of chain deposit will be bridged to.
+        @param recipient Recipient of the deposit.
+        @param  token Input token to be swapped.
+        @param  tokenAmount Amount of tokens to be swapped.
+        @param amountOutMinimum Minimal amount of ETH to be accepted as a swap output.
+        @param pathTokens Addresses of the tokens for Uniswap swap. WETH address is used for ETH.
+        @param pathFees Fees for Uniswap pools.
+    */
+    function depositTokensToEth(
+        uint8 destinationDomainID,
+        address recipient,
+        address token,
+        uint256 tokenAmount,
+        uint256 amountOutMinimum,
+        address[] calldata pathTokens,
+        uint24[] calldata pathFees
+    ) external {
```

Comment:
Shouldn't this be `payable`?

Reply:
In this case what should be done with msg.value? Should it be just added to swapped amount of eth and sent forward to native adapter?

Reply:
Actually, my initial comment was just my misunderstanding. But now that you mentioned, does it make sense to also add this as a feature (as essentially you could then use combination of ETH + ERC20 token even) cc: @mpetrunic

Reply:
We dont have that supported in algo so not mission critical. I would assume later on this would be covered by ass multicall

Reply:
Yep, @mpetrunic is correct about this. There is no need to add this functionality here, so we will close the discussion. cc @viatrix